datastore/longhorn.go

```diff
@@ -585,29 +585,73 @@ func checkEngine(engine *longhorn.Engine) error {
 	return nil
 }
 
-// PickVolumeCurrentEngine pick the current Engine from the Engine list of a volume with the given namespace
-func (s *DataStore) PickVolumeCurrentEngine(v *longhorn.Volume, es map[string]*longhorn.Engine) (*longhorn.Engine, error) {
-	if len(es) == 0 {
-		return nil, nil
+// GetCurrentEngineAndExtras pick the current Engine and extra Engines from the Engine list of a volume with the given namespace
+func GetCurrentEngineAndExtras(v *longhorn.Volume, es map[string]*longhorn.Engine) (currentEngine *longhorn.Engine, extras []*longhorn.Engine, err error) {
+	if len(es) == 1 {
+		for _, e := range es {
+			return e, nil, nil
+		}
 	}
+	for _, e := range es {
+		if e.Spec.Active {
+			currentEngine = e
+		} else {
+			extras = append(extras, e)
+		}
+	}
+	if currentEngine == nil {
+		logrus.Warnf("failed to directly pick up the current one from multiple engines for volume %v, fall back to detect the new current engine, "+
+			"current node %v, desire node %v", v.Name, v.Status.CurrentNodeID, v.Spec.NodeID)
+		return GetNewCurrentEngineAndExtras(v, es)
+	}
+	return
+}
 
+// GetNewCurrentEngineAndExtras detects the new current Engine and extra Engines from the Engine list of a volume with the given namespace during engine switching.
+func GetNewCurrentEngineAndExtras(v *longhorn.Volume, es map[string]*longhorn.Engine) (currentEngine *longhorn.Engine, extras []*longhorn.Engine, err error) {
 	if len(es) == 1 {
 		for _, e := range es {
-			return e, nil
+			return e, nil, nil
 		}
 	}
 
-	// len(es) > 1
-	node := v.Status.CurrentNodeID
-	if node != "" {
-		for _, e := range es {
-			if e.Spec.NodeID == node {
-				return e, nil
-			}
+	oldEngineName := ""
+	for name := range es {
+		e := es[name]
+		if e.Spec.Active {
```

Comment:
What happens if there are multiple active engines should we overwrite? Or is there another condition we can use in that case? i.e first come first server?